website/source/docs/internals/providers/new-data-source.html.md

```diff
@@ -0,0 +1,90 @@
+---
+layout: "docs"
+page_title: "Creating Data Sources"
+sidebar_current: "docs-internals-provider-guide-new-data-source"
+description: |-
+  How to get started adding a data source to an existing provider.
+---
+
+# Creating Data Sources
+
+Once we have a Provider, we can add some Data Sources to that provider, which
+will help it pull information out of your infrastructure to be used in your
+Resources.  To do this, you need to write some code in the package your
+Provider is defined in.
+
+In the built-in plugins for Terraform that are shipped as part of the codebase,
+we have a file naming guideline: we use
+`data_source_{api}_{data_source_name}.go` as the template to name our files,
+and tend to stick to  single Data Source per file. For example, AWS has
+`data_source_aws_acm_certificate.go`, because it reads certificates from AWS’
+ACM service. Some providers only have a single API, so we use the provider name
+in that case. For example, Docker has `data_source_docker_registry_image.go`.
+If you want to contribute your code back to the Terraform repo, it makes things
+easier if you follow this convention.
+
+Inside this file, you’re going to define a function that takes no arguments and
+returns a `*schema.Resource`, just like Resources.  While you can technically
+name your function anything, we recommend (especially if you plan on
+contributing your code back to the Terraform repo!) that you name it the
+camelCase version of the file name. For example, `data_source_aws_eip.go` would
+contain `dataSourceAwsEip`. We’ll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Data Source  With the Provider
+
+Now that you’ve got your function, switch back to your Provider definition, and
+add the Data Source to the Provider’s `DataSourcesMap` property. The key will
+be the name of the Data Source used in state and configuration files, and (by
+convention) should match the `{data_source_name}` part used in the filename for
+the file that defines the Resource. Nothing will necessarily break if you don’t
+do this, but if you happen to accidentally have a collision with any other
+resource in any other provider, things will likely break. This convention helps
+keep things neat and orderly while avoiding conflicts.
+
+## Defining the Data Source Properties
+
+Each Data Source has a set of properties called the “schema” that is stored in
```

Comment:
Fancy quotes. Fancy quotes. Fancy quotes. Please do a mass find-and-replace. These will make the builder sad and is a by-product of working in a Google doc. You can disable fancy quotes in your google settings.